### PR TITLE
Haar 847 - UI email domains crd and integration to the api endpoints

### PR DESCRIPTION
**Pull request by olatomoloju**
Hello, Could you help do the PR please, it covers the email domain create, delete and listing functionality.
Some test were updated, but new cypress and other relevant test will be part of another PR.

### CODE REVIEW

File: backend/routes.js [hunk 1]
```diff
@@ -30,6 +33,13 @@ const configureRoutes = ({ oauthApi, manageUsersApi, nomisUsersAndRolesApi }) =>
   router.use('/manage-dps-users/:userId', manageDpsUserRouter({ oauthApi, nomisUsersAndRolesApi, manageUsersApi }))
   router.use('/manage-groups', manageGroupsRouter({ oauthApi, manageUsersApi }))
   router.use('/manage-roles', manageRolesRouter({ oauthApi, manageUsersApi }))
+  router.use('/email-domains', manageEmailDomainsRouter({ oauthApi, manageUsersApi }))
+  router.use('/create-email-domain', createEmailDomainsRouter({ oauthApi, manageUsersApi }))
+  router.use(
+    '/delete-email-domain?id=:domainId&name=:domainName',
+    deleteEmailDomainsRouter({ oauthApi, manageUsersApi }),
+  )
+  router.use('/delete-email-domain', deleteEmailDomainsRouter({ oauthApi, manageUsersApi }))
```
Comment (karenmillermoj) [inline backend/routes.js:42]: Do these need oauthApi?

File: backend/routes/createEmailDomainRouter.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  const { index: createEmailIndex, post: createEmailExec } = createEmailDomainFactory(
+    createEmailDomainApi,
+    '/create-email-domain',
+    '/email-domains',
+  )
+
+  router.get('/', createEmailIndex)
+  router.post('/', createEmailExec)
+
```
Comment (karenmillermoj) [inline backend/routes/createEmailDomainRouter.js:16]: naming is a bit confusing, possibly createEmailDomainXX

File: backend/views/pages/email-domains/allEmailDomains.njk [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+{% from "govuk/components/input/macro.njk" import govukInput %}
+
+{% set pageTitle = applicationName + " - Court Register" %}
+
+{% set filterOptionsHtml %}
+  <form id="filter-form" action="/court-register" novalidate>
+    {{ govukInput( allCourtsFilter | toCourtTextSearchInput ) }}
+    {{ govukRadios( allCourtsFilter | toCourtActiveFilterRadioButtons ) }}
+    {{ govukCheckboxes( courtTypes | toCourtTypeFilterCheckboxes(allCourtsFilter) ) }}
+  </form>
```
Comment (karenmillermoj) [inline backend/views/pages/email-domains/allEmailDomains.njk:19]: Are we filtering? and this looks like court code

File: backend/views/pages/email-domains/allEmailDomains.njk [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  <div class="govuk-width-container">
+    {{ govukBackLink({
+      text: "Back",
+      href: "/"
+    }) }}
+
+    <main class="app-container govuk-body">
+      <div class="govuk-grid-row">
+        <div class="govuk-grid-column-full">
+          <h1 class="govuk-heading-xl">Court Register</h1>
```
Comment (karenmillermoj) [inline backend/views/pages/email-domains/allEmailDomains.njk:33]: Court?

File: backend/views/pages/email-domains/allEmailDomains.njk [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        </div>
+      </div>
+
+      <h1>Helloooooo</h1>
+    </main>
```
Comment (karenmillermoj) [inline backend/views/pages/email-domains/allEmailDomains.njk:38]: invalid data

Comment (karenmillermoj) [reply backend/views/pages/email-domains/allEmailDomains.njk:38]: I can't see where the list is shown

File: backend/views/pages/email-domains/allEmailDomains.njk [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  </div>
+
+{% endblock %}
+
+{% block bodyEnd %}
+  {{ super() }}
+  <script src="/assets/filter-toggle-button.js"></script>
+{% endblock %}
```
Comment (karenmillermoj) [inline backend/views/pages/email-domains/allEmailDomains.njk:46]: filtering not required

File: backend/views/pages/email-domains/allEmailDomains.njk (hunk 5 above)
Comment (olatomoloju) [reply backend/views/pages/email-domains/allEmailDomains.njk:38]: Some cleanup required.

File: backend/views/pages/email-domains/allEmailDomains.njk (hunk 4 above)
Comment (olatomoloju) [reply backend/views/pages/email-domains/allEmailDomains.njk:33]: Cleanup required

File: backend/views/pages/email-domains/allEmailDomains.njk (hunk 3 above)
Comment (olatomoloju) [reply backend/views/pages/email-domains/allEmailDomains.njk:19]: Code littering from the original approach, now deleted.

File: backend/views/pages/email-domains/allEmailDomains.njk (hunk 4 above)
Comment (olatomoloju) [reply backend/views/pages/email-domains/allEmailDomains.njk:33]: Same as above.

File: views/deleteEmailDomain.njk [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      {% if hasMaintainAuthUsers %}
+        {{ govukButton({
+          text: "Delete Email Domain",
+          type: "submit",
+          classes: "govuk-button--warning",
+          attributes: { "data-qa": "delete-domain-button" }
+        }) }}
+      {% else %}
+        <p>You do not have permission to delete this group</p>
+      {% endif %}
```
Comment (karenmillermoj) [inline views/deleteEmailDomain.njk:45]: authorisationMiddleware.ts should store this logic

File: views/emailDomainListing.njk [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                            <td class="govuk-table__cell">{{ domain.domain }}</td>
+                            <td class="govuk-table__cell">{{ domain.description }}</td>
+                            <td class="govuk-table__cell"><a href="/delete-email-domain?id={{ domain.id }}&name={{ domain.domain }}"
+                                 class="govuk-link govuk-link--no-visited-state">Delete</a>
+                            </td>
+                          </tr>
+                        {% endfor %}
+                        </tbody>
+                      </table>
+                    {% else %}
```
Comment (karenmillermoj) [inline views/emailDomainListing.njk:70]: trailing else block

File: backend/routes/createEmailDomainRouter.js (hunk 2 above)
Comment (olatomoloju) [reply backend/routes/createEmailDomainRouter.js]: Done

File: backend/views/pages/email-domains/allEmailDomains.njk (hunk 6 above)
Comment (olatomoloju) [reply backend/views/pages/email-domains/allEmailDomains.njk]: As above

File: backend/routes.js (hunk 1 above)
Comment (olatomoloju) [reply backend/routes.js]: As discussed, pattern predates this PR, but changes have been done retrospectively.

File: views/emailDomainListing.njk (hunk 8 above)
Comment (olatomoloju) [reply views/emailDomainListing.njk]: done

File: backend/api/manageUsersApi.js [hunk 9]
```diff
@@ -20,6 +20,11 @@ const manageUsersApiFactory = (client) => {
 
   const createUser = (context, user) => post(context, '/prisonusers', user)
   const contextUserRoles = (context, username) => get(context, `/prisonusers/${username}/roles`)
+  const getAllEmailDomains = (context) => {
+    return client.get(context, `/email-domains`).then(processPageResponse(context))
+  }
```
Comment (karenmillermoj) [inline backend/api/manageUsersApi.js:25]: Your call to get all emaildomains is not a paged response. I may have missed this in the first review

File: backend/controllers/addEmailDomain.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+      errors: req.flash('createEmailDomainErrors'),
+    })
+  }
+
+  const post = async (req, res) => {
+    const domain = trimObjValues(req.body)
+    domain.domainDescription = domain.domainDescription.toUpperCase()
+    const errors = validateCreateDomain(domain)
+
+    // const { domainName, domainDescription } = req.body
```
Comment (karenmillermoj) [inline backend/controllers/addEmailDomain.js:27]: comment can be removed

File: backend/routes.js [hunk 11]
```diff
@@ -29,7 +32,11 @@ const configureRoutes = ({ oauthApi, manageUsersApi, nomisUsersAndRolesApi }) =>
   router.use('/manage-external-users/:userId', manageAuthUserRouter({ oauthApi, manageUsersApi }))
   router.use('/manage-dps-users/:userId', manageDpsUserRouter({ oauthApi, nomisUsersAndRolesApi, manageUsersApi }))
   router.use('/manage-groups', manageGroupsRouter({ manageUsersApi }))
-  router.use('/manage-roles', manageRolesRouter({ oauthApi, manageUsersApi }))
+  router.use('/manage-roles', manageRolesRouter({ manageUsersApi }))
+  router.use('/email-domains', manageEmailDomainsRouter({ manageUsersApi }))
+  router.use('/create-email-domain', createEmailDomainsRouter({ manageUsersApi }))
+  router.use('/delete-email-domain?id=:domainId&name=:domainName', deleteEmailDomainsRouter({ manageUsersApi }))
+  router.use('/delete-email-domain', deleteEmailDomainsRouter({ manageUsersApi }))
```
Comment (karenmillermoj) [inline backend/routes.js:40]: Do we need both for delete or is the router clever enough to know they're the same (I don't know without testing)

**Review by karenmillermoj — APPROVED** [review]
lgtm

File: backend/api/manageUsersApi.js (hunk 9 above)
Comment (olatomoloju) [reply backend/api/manageUsersApi.js:25]: Nope, the page response is generic enough to handle non paged calls.

File: backend/routes.js (hunk 11 above)
Comment (olatomoloju) [reply backend/routes.js:40]: Yes, I do need both.

File: backend/controllers/addEmailDomain.js (hunk 10 above)
Comment (olatomoloju) [reply backend/controllers/addEmailDomain.js]: Yes, done.

**Pull request merged**